Automattic/jetpack
# Require plugins to filter available modules to be able to activate/deactivate it

This PR introduces a way for us to keep the return of `Modules::get_active()` consistent with the list of available modules depending on the plugins that are active.

While we keep the option intact, we will only return the active modules that are available at the moment. With that, we ensure the Sync `get_active_modules` callable will only sync modules that are really active - while we can keep the option holding modules that will be active again once Jetpack plugin is re-activated.

#### Changes proposed in this Pull Request:

* Adds a filter to allow plugins to modify the list of available modules when Jetpack is not there
* When getting the list of active plugins, only return those that are also available

#### Jetpack product discussion


p1652965539642009-slack-CBG1CP4EN

#### Does this pull request change what data or activity we track or use?


No

#### Testing instructions:






* Activate Jetpack and connect
* Activate Search
* Deactivate Jetpack
* `wp shell`
* `$x = new Automattic\Jetpack\Modules();`
* `$x->get_available();` should return only search
* get_option( 'jetpack_active_modules' )` should also return jetpack modules
* `$x->activate( 'search', false, false )` should return true
* `$x->get_active` should only return search

Activate social and see social entering the list

Reactivate Jetpack and see all jetpack modules back to the list
